internal/controller/deletion/deletion.go

```diff
@@ -0,0 +1,256 @@
+/*
+Copyright 2024.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package deletion
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"go.uber.org/zap"
+
+	"k8s.io/apimachinery/pkg/api/equality"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/utils/ptr"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/stolostron/siteconfig/api/v1alpha1"
+	ci "github.com/stolostron/siteconfig/internal/controller/clusterinstance"
+	"github.com/stolostron/siteconfig/internal/controller/conditions"
+	cierrors "github.com/stolostron/siteconfig/internal/controller/errors"
+)
+
+const DefaultDeletionTimeout = 30 * time.Minute
+
+type DeletionContext struct {
+	Context         context.Context
+	Client          client.Client
+	Logger          *zap.Logger
+	TimeoutDuration *time.Duration
+}
+
+// NewDeletionContext creates a new DeletionContext with default DeletionTimeout if unspecified.
+func NewDeletionContext(
+	ctx context.Context,
+	client client.Client,
+	logger *zap.Logger,
+	timeoutDuration *time.Duration,
+) *DeletionContext {
+	// Use a default timeout if none is provided
+	if timeoutDuration == nil {
+		timeoutDuration = ptr.To(DefaultDeletionTimeout)
+	}
+
+	if ctx == nil || client == nil || logger == nil {
+		return nil
```

Comment:
This should return an error for what is effectively a parameter validation failure.

Reply:
This is now removed.